app/Http/Controllers/App/UsersController.php

```diff
@@ -36,58 +48,176 @@ public function add(Request $request)
         $user->fill($request->input('user'));
+
+        if (!$request->has('user.password_new')) {
+            $user->save();
+            DB::commit();
+
+            return self::json($user->toArrayCamelize(), 200);
+        }
+
+        if ($token_authorization) {
+            $user->password = $request->input('user.password_new');
+            $user->save();
+            DB::commit();
+
+            return self::json($user->toArrayCamelize(), 200);
+        }
+
+        if (!$request->has('user.password_old') || !$user->validPassword($request->input('user.password_old'))) {
+            return self::json($user->toArrayCamelize(), 422, ['password_old' => 'Old password is not valid']);
+        }
+
+        $user->password = $request->input('user.password_new');
         $user->save();
+        DB::commit();
 
         return self::json($user->toArrayCamelize(), 200);
     }
 
     public function emailActivate(Request $request)
     {
-        $this->validateRequest($request, 'user', User::$rules_email_activate);
-
-        $user = User::where(
-            'email_confirm_token',
-            $request->input('user.email_confirm_token')
-        )->whereNull('email_confirmed_at')->first();
+        Validator::make($request->all(), ['user.email_confirm_token' => 'required'])->validate();
 
+        DB::beginTransaction();
+        if (false === $token = Token::check($request->input('user.email_confirm_token'))) {
+            return self::json([], 403);
+        }
+        $user = User::find($token['user_id']);
         if (empty($user)) {
-            return self::json([], 401);
+            return self::json([], 403);
         }
-
         $user->email_confirmed_at = date('Y-m-d H:i:s');
         $user->save();
+        DB::commit();
+
+        return self::json($user->toArrayCamelize(), 200);
+    }
+
+    public function emailActivateResend(Request $request)
+    {
+        Validator::make($request->all(), ['uri' => 'required'])->validate();
+        $user = Auth::user();
+        DB::beginTransaction();
+        if (!Token::canGenerate($user->id, 'email-activate', $this->email_activation_resend_limit)) {
+            return self::json([], 400, ['message' => 'You can request 1 email activation every 15 minutes. Please wait.']);
+        }
+        Mail::to($user)->queue(new UserEmailActivate(
+            Token::generate('email-activate', $this->email_activation_token_time, $user->id),
+            $request->input('uri')
+        ));
+        DB::commit();
+
+        return self::json([], 204);
+    }
+
+    public function emailChangeStep1(Request $request)
+    {
+        Validator::make($request->all(), ['email' => 'required|email', 'URIstep1' => 'required', 'URIstep2' => 'required'])->validate();
+        if (User::withTrashed()->where('email', $request->input('email'))->count()) {
+            return self::json([], 422, ['email' => 'This email already exists in our database']);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "This email already exists in our database" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)